src/nvim/api/vim.c

```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
   return colors;
 }
 
+// Helper function for nvim_get_keymap and nvim_buf_get_keymap
+// @param mode The abbreviation for the mode
+// @param buffer_id
+//      -1: Global mapping
+//      0: Current buffer
+//      else: The desired buffer ID
+//
+// @returns An array of maparg() like dictionaries describing mappings
+static Array get_keymap_helper(String mode, Integer buffer_id)
+{
+  Array mappings = ARRAY_DICT_INIT;
+  dict_T *const dict = dict_alloc();
+
+  // Convert the string mode to the integer mode
+  // that is stored within each mapblock
+  int int_mode = get_map_mode((char_u **)(&(mode.data)), 0);
+
+  for (int i = 0; i < MAX_MAPHASH; i++) {
+    for (mapblock_T *current_maphash = get_maphash(i, (long)buffer_id);
+        current_maphash;
+        current_maphash = current_maphash->m_next) {
+      // Check for correct mode
+      if ((int_mode & current_maphash->m_mode) != 0) {
+        mapblock_fill_dict(
+            dict,
+            current_maphash,
+            (long)buffer_id,
+            false);
+
+        ADD(mappings,
+            vim_to_object((typval_T[]) {
+                { .v_type = VAR_DICT, .vval.v_dict = dict }
+              })
+            );
+
+        dict_clear(dict);
+      }
+    }
+  }
+  dict_free(dict);
+
+  return mappings;
+}
+
+// Get a list of dictionaries describing global (i.e. non-buffer) mappings
+// @param mode The abbreviation for the mode
+// @returns An array of maparg() like dictionaries describing mappings
+Array nvim_get_keymap(String mode)
+  FUNC_API_SINCE(2)
+{
+  return get_keymap_helper(mode, -1);
+}
+
+
+// Get a lit of dictionaries describing buffer-local mappings
+// @param mode The abbreviation for the mode
+// @param buffer_id
+//      0: Current buffer
+//      else: The desired buffer ID
+//
+// @returns An array of maparg() like dictionaries describing mappings
+Array nvim_buf_get_keymap(String mode, Integer buffer_id)
```

Comment:
This should be in `api/buffer.c`, also first argument should be `Buffer buffer`.